My Site Dashboard: Custom feature announcement

hassaanelgarem opened the pull request:
Closes #18209

## Description
- Refactored `WhatIsNewView` to depend on an appearance object that gets injected when creating the view.
- Created a new announcement cell to match the new design
- Used `appVersionName` to determine if the announcement is for the dashboard
- Extracted the view's title from the returned features. 
- Injected custom appearance and new custom cell when showing announcement for dashboard  

/ | Light | Dark
-- | -- | --
Dashboard | <img src="https://user-images.githubusercontent.com/25306722/160506159-924b1970-052b-4eed-806f-e41c7158d926.png"> | <img src="https://user-images.githubusercontent.com/25306722/160506175-e651abcd-e88e-41e2-8b64-85c3ad0dd567.png">
Default | <img src="https://user-images.githubusercontent.com/25306722/160507808-a950b9ee-6d26-41c5-beb2-b0d39b40d38c.png"> | <img src="https://user-images.githubusercontent.com/25306722/160507787-5e3e5900-69e4-4d1c-84e3-0df534578168.png">

## Testing Instructions
First you need to modify the response of the feature announcement API to return custom response for testing. This can be achieved in multiple ways. Here are two options:

<details>
  <summary><strong>Hardcoding response in code (Easier)</strong></summary>
  <br>

  - Replace `AnnouncementServiceRemote.swift` in WordPressKit with this [version](https://drive.google.com/file/d/1HUtBzTB7vdDjNhwXaAMPZERBD9JX0rZ9/view?usp=sharing) (This file is in an external pod. Make sure to revert the changes after you're done by removing the pod from Pods folder and running `bundle exec pod install`)
  - Replace `AnnouncementsStore.swift` by this [version](https://drive.google.com/file/d/1RZ4xae57r3XFnHo-kd5ciaX9PO_l1ehj/view?usp=sharing)
  ---
</details>
<details>
  <summary><strong>Using Proxy Apps</strong></summary>
  <br>

  - Install [ProxyMan](https://proxyman.io)
  - Configure ProxyMan to [capture simulator traffic](https://docs.proxyman.io/debug-devices/ios-simulator)   
  - Make sure the app's network traffic shows up in ProxyMan
  - Add a local map for the targetted API
      - Go to Tools -> Map Local...
      - Add new mapping
      - Set URL to `https://public-api.wordpress.com/wpcom/v2/mobile/feature-announcements`
      - Make sure to check "Include all subpaths of this URL"
      - Under "Map to" choose Local Directory and use this [file](https://drive.google.com/file/d/1McI4p5Gl-Ew59Ql7mngpwSPoLVp9PRyh/view?usp=sharing)
  ---
</details>

_PS: Once I have access to the Mobile Announce mc tool, I will create an announcement targeting this PR, which should remove the need for the above requirement._

1. Make sure to follow one of the above options
2. To ease the testing modify `WhatIsNewScenePresenter.shouldPresentWhatIsNew` to always return true
3. Run the app
4. The announcement should be displayed and should match the design
5. Change the `appVersionName` in the response to anything that's not "19.6"
6. Rerun the app
7. The announcement should be displayed and should match the original design
8. Go to My Site
9. Tap the FAB then tap Story Post
10. Make sure Stories intro is displayed correctly
11. Please test light and dark mode

## Regression Notes
1. Potential unintended areas of impact
Normal feature announcement view and stories intro

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually the showing of normal announcements

12. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hassaanelgarem committed:
Refactor: make `WhatIsNewView.Appearance` a reusable, injectable struct

hassaanelgarem committed:
Add: choose between standard and custom what's new view

hassaanelgarem committed:
Add: customize strings for custom msd announcement

hassaanelgarem committed:
Add: configure dashboard feature announcement cell

hassaanelgarem committed:
Modify: configure custom dashboard announcement appearance

hassaanelgarem committed:
Add: disclaimer label to dashboard feature announcement

hassaanelgarem committed:
Update: dashboard announcement cell estimated row height

hassaanelgarem commented:
@osullivanchris Let me know if you have any comments!

And regarding the button font, I kept the font used by the normal announcement which is system font medium 22pt.

osullivanchris commented:
@hassaanelgarem thanks for the ping, it looks fantastic 🎉 

Two minor nitpicks:
- Could we add another 8pt padding to the left and right of the text? It runs a little long for the first paragraph. Looks too close to the edge of the screen. I know this means its not technically aligned precisely to the edge of the image but I think it will look better.
- The second image is floating a little bit above the bottom edge of the background rectangle, looks like it needs to be nudged down slightly

> and regarding the button font, I kept the font used by the normal announcement which is system font medium 22pt.

Ah yes perfect. I knew there was an existing one we could pull from. Looks fine.

osullivanchris commented:
Added 'Needs Copy Review' label.

FAO Editorial:
- this is an introduction to the new dashboard we are launching. 
- The first point I am trying to get across is that theres a new space with dynamic content about the users site. Its a fresh, up to date space to get a summary.
- In the second point I am trying to reassure them that the menu they rely on isn't going away. We are introducing tabs so the user can have the utility of the list, or the vibrancy of the dashboard

I think my copy has mostly covered what I'm trying to say. But would love a closer eye and any suggestions.

hassaanelgarem committed:
Change: add horizontal padding to heading label in DashboardCustomAnnouncementCell

hassaanelgarem commented:
@osullivanchris 
> Could we add another 8pt padding to the left and right of the text? It runs a little long for the first paragraph. Looks too close to the edge of the screen. I know this means its not technically aligned precisely to the edge of the image but I think it will look better.

I agree, it does look better! Let me know if you want to increase the padding further.

> The second image is floating a little bit above the bottom edge of the background rectangle, looks like it needs to be nudged down slightly

I've addressed this by adjusting the image used. I added some padding to the top of the image to match the size of the other image. This is the [new image](https://user-images.githubusercontent.com/25306722/160597552-38dccea9-dd27-43d7-98a6-6e019b62e74a.png) I'm using

This is how it looks after applying these changes:

<img src="https://user-images.githubusercontent.com/25306722/160599805-6e54f15d-efd9-44cb-8e96-20d5bfd0867d.png" width=50%>

osullivanchris commented:
> I agree, it does look better! Let me know if you want to increase the padding further.

that looks good thanks!

> I've addressed this by adjusting the image used. I added some padding to the top of the image to match the size of the other image. This is the [new image](https://user-images.githubusercontent.com/25306722/160597552-38dccea9-dd27-43d7-98a6-6e019b62e74a.png) I'm using

aha! I didn't think of that. When one image had some blank space. I could have accounted for that. It looks good with your adjustment, thanks.

Happy with both those fixes 👍

leandroalonso commented:
@osullivanchris @hassaanelgarem 

Not necessarily related to this PR but the screen on iPad is biiig. If we change the `WhatIsNewViewController` to be `modalPresentationStyle = .formSheet` (you can do that on `init`) the result looks better IMHO:

| Current | Form Sheet |
| ------- | ------------ |
| ![Simulator Screen Shot - iPad Pro (9 7-inch) - 2022-03-29 at 09 49 38](https://user-images.githubusercontent.com/7040243/160616079-b8ca4635-d2a8-43f2-ab6a-01a994569304.png) | ![Simulator Screen Shot - iPad Pro (9 7-inch) - 2022-03-29 at 09 54 34](https://user-images.githubusercontent.com/7040243/160616093-c01da2f4-f749-4b52-b89d-814382cac6f3.png) |

leandroalonso commented:
@hassaanelgarem it worked fine with the JSON you provided.

However, I went ahead and configured an announcement test. This is [the JSON](https://gist.github.com/leandroalonso/da0b750a6984332c3a8e061a71f1d684), and this is how it looked in the tool:

<img width="1263" alt="Screen Shot 2022-03-29 at 10 50 49" src="https://user-images.githubusercontent.com/7040243/160626886-017956b0-9a31-4b65-b2b4-66cdd1a90370.png">

For some reason that was the result on the app:

<img src="https://user-images.githubusercontent.com/7040243/160626996-5bd66948-73d6-4c6e-93ec-9c349f746dc4.png" width="300">

Edit: I believe this is because of the `"."` on `title` and `subtitle`. The tool doesn't allow to create an announcement without those, so we need to add something.

osullivanchris commented:
> Not necessarily related to this PR but the screen on iPad is biiig. If we change the WhatIsNewViewController to be modalPresentationStyle = .formSheet (you can do that on init) the result looks better IMHO:

Agree. I'd like to go with that.

hassaanelgarem committed:
Change: modal presentation style of `WhatIsNewVC` to formSheet

hassaanelgarem committed:
Change: Treat feature strings of value "." as empty strings

hassaanelgarem commented:
@leandroalonso 

> If we change the WhatIsNewViewController to be modalPresentationStyle = .formSheet (you can do that on init) the result looks better

Done in b778b0fbcd1fef4cc31772b694869a78e8bbb4a2

> I believe this is because of the "." on title and subtitle. The tool doesn't allow to create an announcement without those, so we need to add something.

I depended on titles or subtitles being empty. Modified the code in 0007112d501f23a43421b0b917c6dda2c153da9a to handle "." as empty strings. It should work as expected now 👍

clucasrowlands commented:
> I think my copy has mostly covered what I'm trying to say. But would love a closer eye and any suggestions.

Hi! I think your proposed copy is pretty clear and succinct! But here are a few slight revisions to your current wording, plus a few additional options, if you'd like to consider.

**First line:**

> My Site has a new home tab that shows you up-to-date highlights from your site.

> My Site has a new home tab that shows you the latest highlights from your site.

**Second line:**

> Your menu hasn't gone anywhere. You can switch between your home and menu tabs any time.

> You can still access your menu! Switch between your home and menu tabs any time.

> You can switch between this new home tab and your menu any time.

leandroalonso reviewed (approved):
Tested with a real announcement on the tool, works as expected!

osullivanchris commented:
@clucasrowlands thanks for taking a look. Those options work great. Let's go with

> My Site has a new home tab that shows you up-to-date highlights from your site.

> You can still access your menu! Switch between your home and menu tabs any time.

Here's how it looks with the final copy

cc @hassaanelgarem 

![ios custom announcement final copy](https://user-images.githubusercontent.com/50576199/160798154-3ad113bf-4379-46e7-ad04-0cd185be4a5a.png)
